amethyst_renderer/src/mesh.rs

```diff
@@ -1,44 +1,181 @@
+
+    /// Get vertex count in buffer
+    fn len(&self) -> usize;
+
+    /// Build `VertexBuffer`
+    fn build(&self, factory: &mut Factory) -> Result<VertexBuffer>;
+}
+
+/// Construct new vertex data from raw data and vertex format
+pub fn vertex_data<D, V>(data: D) -> (D, PhantomData<V>)
+where
+    D: AsRef<[V]>,
+    V: VertexFormat,
+{
+    (data, PhantomData)
+}
+
+impl<D, V> VertexData for (D, PhantomData<V>)
+where
+    D: AsRef<[V]>,
+    V: VertexFormat,
+{
+    const ATTRIBUTES: Attributes<'static> = V::ATTRIBUTES;
+
+    fn len(&self) -> usize {
+        self.0.as_ref().len()
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<VertexBuffer> {
+        use gfx::Factory;
+        use gfx::memory::{cast_slice, Bind};
+        use gfx::buffer::Role;
+
+        let verts = self.0.as_ref();
+        let slice = cast_slice(verts);
+        let stride = slice.len() / verts.len();
+        let role = Role::Vertex;
+        let bind = Bind::empty();
+
+        let vbuf = factory.create_buffer_immutable_raw(slice, stride, role, bind)?;
+        Ok(VertexBuffer {
+            attrs: V::ATTRIBUTES,
+            raw: vbuf,
+        })
+    }
+}
+
+/// Set of vertex data
+#[doc(hidden)]
+pub trait VertexDataSet {
+    /// Iterator for `VertexBuffer`s built
+    type VertexBufferIter: Iterator<Item = VertexBuffer>;
+
+    /// Get smalles vertex count across buffers
+    fn len(&self) -> usize;
+
+    /// Build `VertexBuffer`s
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter>;
+}
+
+impl<H> VertexDataSet for (H, ())
+where
+    H: VertexData,
+{
+    type VertexBufferIter = Once<VertexBuffer>;
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter> {
+        let (ref head, _) = *self;
+        Ok(once(head.build(factory)?))
+    }
+}
+
+impl<H, T> VertexDataSet for (H, T)
+where
+    H: VertexData,
+    T: VertexDataSet,
+{
+    type VertexBufferIter = Chain<Once<VertexBuffer>, T::VertexBufferIter>;
+
+    fn len(&self) -> usize {
+        use std::cmp::min;
+        min(self.0.len(), self.1.len())
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter> {
+        let (ref head, ref tail) = *self;
+        Ok(once(head.build(factory)?).chain(tail.build(factory)?))
+    }
+}
+
 
 /// Represents a polygonal mesh.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug)]
 pub struct Mesh {
-    attrs: &'static [(&'static str, AttributeFormat)],
-    prim: Primitive,
     slice: Slice,
     transform: Matrix4<f32>,
-    vbuf: RawBuffer,
+    vbufs: SmallVec<[VertexBuffer; 4]>,
 }
 
 impl Mesh {
     /// Builds a new mesh from the given vertices.
-    pub fn build<D, V>(verts: D) -> MeshBuilder<D, V>
-    where
-        D: AsRef<[V]>,
-        V: VertexFormat,
+    pub fn build<D, V>(verts: D) -> MeshBuilder<((D, PhantomData<V>), ())>
+        where D: AsRef<[V]>,
```

Comment:
This is definitely not the new rustfmt style

Reply:
Add `--all` att the end

Reply:
I like old format :frowning: